### PR TITLE
Implement bep3 evm native conversion logic

### PR DESCRIPTION
**Pull request by DracoLi**
## Checklist
 - [ ] Changelog has been updated as necessary.

### CODE REVIEW

File: x/evmutil/keeper/conversion_evm_native_bep3.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package keeper
+
+import (
+	"math/big"
+
+	"github.com/kava-labs/kava/x/evmutil/types"
+)
+
+var (
+	DefaultBEP3ConversionDenoms = []string{"bnb", "busd", "btcb", "xrpb"}
```
Comment (rhuairahrighairidh) [inline x/evmutil/keeper/conversion_evm_native_bep3.go:10]: nit, I'd make this variable private (or use constants) to avoid other packages being able to modify it. Low chance of it being a problem but adding unauthorised assets here would allow minting of coins

File: x/evmutil/keeper/conversion_evm_native.go [hunk 2]
```diff
@@ -102,13 +110,32 @@ func (k Keeper) ConvertERC20ToCoin(
...
+	// When converting bep3 erc20 assets to sdk.Coin, the following will happen
+	// 1. Convert provided 18 decimals erc20 amount to 8 decimals sdk amount, then mint those.
+	//    Dust from converting 18 decimals to 8 decimals are ignored and will
+	//    remain in the original wallet.
+	// 2. Convert to-be-minted 8 decimals sdk.Coin back to 18 decimals, then lock those.
+	if IsEvmNativeBep3Conversion(pair) {
+		amountToMint, _ = ConvertBep3ERC20AmountToCoinAmount(amount.BigInt())
+
+		// make sure we have at least 1 sdk.Coin to mint
+		if amountToMint.Cmp(big.NewInt(0)) == 0 {
```
Comment (rhuairahrighairidh) [inline x/evmutil/keeper/conversion_evm_native.go:126]: good catch on this edge case, would be good to cover in a test (in this or levi's branch)

**Review by rhuairahrighairidh — COMMENTED** [review]
looks good

File: x/evmutil/keeper/conversion_evm_native.go [hunk 3]
```diff
@@ -71,7 +71,15 @@ func (k Keeper) ConvertCoinToERC20(
 		return err
 	}
 
-	if err := k.UnlockERC20Tokens(ctx, pair, coin.Amount.BigInt(), receiverAccount); err != nil {
+	// handle bep3 conversion pair (8 decimals sdk.Coin -> 18 decimals erc20)
+	// when converting bep3 bank coins to erc20, we will need to unlock the 18
+	// decimals erc20 equivalent of the 8 decimals bep3 sdk.Coins
+	amountToUnlock := coin.Amount.BigInt()
+	if IsEvmNativeBep3Conversion(pair) {
```
Comment (nddeluca) [inline x/evmutil/keeper/conversion_evm_native.go:78]: After a map is used & conversion factor is pulled to a top level variable, I wonder if 3 private methods are necessary.

option 1
```golang
// coin -> erc20
amountToUnlock := coin.Amount
if _, isBep3Asset := bep3Denoms[coin.Denom]; isBep3Asset {
  amountToUnlock = amountToUnlock.Mul(bep3ConversionFactor)
}

// erc20 -> coin
amountToLock := amount
amountToMint := amount
if _, isBep3Asset := bep3Denoms[pair.Denom]; isBep3Asset {
  amountToMint = amountToMint.Quo(bep3ConversionFactor)

  if amountToMint.IsZero() {
    return errorsmod.Wrapf(types.ErrInsufficientConversionAmount, "unable to convert bep3 coin due converting less than 1%s", pair.Denom)
  }

  amountToLock := amountToMint.Mul(bep3ConversionFactor)
}
```

vs option 2

```golang
// coin -> erc20
amountToUnlock := coin.Amount
if isBep3Asset(coin.Denom) {
  amountToUnlock = covertBep3CoinToErc20(amountToUnlock)
}

// erc20 -> coin
amountToLock := amount
amountToMint := amount
if isBep3Asset(pair.Denom) {
  amountToMint := convertBep3Erc20ToCoin(amountToMint)

  if amountToMint.IsZero() {
    return errorsmod.Wrapf(types.ErrInsufficientConversionAmount, "unable to convert bep3 coin due converting less than 1%s", pair.Denom)
  }

  amountToLock := convertBep3CoinToErc20(amountToLock)
```

vs option 3

```golang
// coin -> erc20
amountToUnlock := convertToErc20Decimals(coin)

// erc20 -> coin
amountToLock, amountToMint, err := calculateAmountToLockAndMint(pair.Denom, amount)
if err != nil {
  return err
}
```

- Option 1 is the default implementation, but doesn't match the complexity level of the calling function well.
- Option 2 it may use some named functions, but is the complexity of the top level function actually reduced?  Also, is `amountToLock := convertBep3CoinToErc20(amountToLock)` actually more comprehensible than `amountToLock := amountToMint.Mul(bep3ConversionFactor)`?  Enough so that the extra function is worth it?
- Option 3 naming could be better, but this pushes the complexity down a level -- improving readability at the top level, but also keeping the implementation easy to find & understand.  IMO, this seems like the best balance.

**Review by nddeluca — CHANGES_REQUESTED** [review]
Really good test cases covering dust, too small of balance, and insuffient balances on both the sdk and erc20 sides.

Let's just add some defensive testing to ensure all four bep3 assets are covered. In addition to using a map for denom lookup, removing unused code, and making extracted helpers private (or adding unit tests for them).

Looks great otherwise.  Refactoring implementation is optional -- just added my thoughts there on readability & complexity.

File: x/evmutil/keeper/conversion_evm_native_bep3.go (hunk 1 above)
Comment (nddeluca) [reply x/evmutil/keeper/conversion_evm_native_bep3.go]: Good catch @rhuairahrighairidh

File: x/evmutil/keeper/conversion_evm_native.go (hunk 3 above)
Comment (DracoLi) [reply x/evmutil/keeper/conversion_evm_native.go:78]: I do prefer option 2's separation of the bep3 logic with the rest of the conversion logic. It's very clear when reading the conversion functions that bep3 assets are handled differently, and we can easily remove that logic in the future without touching any existing conversion logic. 
I do also think abstracting some of the bep3 logic will help with readability, especially for the erc20 -> coin conversion logic. I'll test out a hybrid of option 2 and 3's approach.

Comment (nddeluca) [reply x/evmutil/keeper/conversion_evm_native.go:78]: Cool 👍. Ya do agree option 3 doesn't clearly show it's for bep3 only, hybrid approach would be good

Comment (DracoLi) [reply x/evmutil/keeper/conversion_evm_native.go]: https://github.com/Kava-Labs/kava/pull/1848/commits/98fc5962aaee656d87004abbf367211b1db70eca

File: x/evmutil/keeper/conversion_evm_native_bep3.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"math/big"
+
+	errorsmod "cosmossdk.io/errors"
+	"github.com/kava-labs/kava/x/evmutil/types"
+)
+
+var (
+	bep3Denoms = map[string]bool{
```
Comment (nddeluca) [inline x/evmutil/keeper/conversion_evm_native_bep3.go:11]: Nice, normally I'd say use a `map[string]struct{}`, but with 4 items the benefited clarify and simpler logic in `isBep3Asset` is nice 👍

File: x/evmutil/keeper/conversion_evm_native_bep3.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	bep3ConversionFactor = new(big.Int).Exp(big.NewInt(10), big.NewInt(10), nil)
+)
+
+func isBep3Asset(denom string) bool {
+	return bep3Denoms[denom]
+}
+
+// convertBep3CoinAmountToERC20Amount converts a bep3 coin amount with 8 decimals
+// to the equivalent ERC20 token with 18 decimals.
+func convertBep3CoinAmountToERC20Amount(amount *big.Int) *big.Int {
```
Comment (nddeluca) [inline x/evmutil/keeper/conversion_evm_native_bep3.go:27]: Just a personal preference and no need to change, but I like named return arguments for functions like this.

Ex
```golang
func convertBep3CoinAmountToERC20Amount(amount *big.Int) (result *big.Int) {
	return new(big.Int).Mul(amount, bep3ConversionFactor)
}
func convertBep3ERC20AmountToCoinAmount(amount *big.Int) (quotient *big.Int) {
	return new(big.Int).Div(amount, bep3ConversionFactor)
}

File: x/evmutil/keeper/conversion_evm_native_bep3.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// convertBep3ERC20AmountToCoinAmount converts a bep3 ERC20 token with 18 decimals
+// to the equivalent coin amount with 8 decimals, and dropping the remainder.
+func convertBep3ERC20AmountToCoinAmount(amount *big.Int) *big.Int {
+	quotient := new(big.Int).Div(amount, bep3ConversionFactor)
+	return quotient
+}
+
+// bep3ERC20AmountToCoinMintAndERC20LockAmount converts 18 decimals erc20 bep3
```
Comment (nddeluca) [inline x/evmutil/keeper/conversion_evm_native_bep3.go:39]: appreciate the doc comments 👍

File: x/evmutil/keeper/conversion_evm_native_bep3_test.go [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+var (
+	bep3Denoms = []string{"bnb", "busd", "btcb", "xrpb"}
+)
+
+func TestBep3ConversionTestSuite(t *testing.T) {
+	suite.Run(t, new(Bep3ConversionTestSuite))
+}
+
+func (suite *Bep3ConversionTestSuite) TestConvertCoinToERC20_Bep3() {
+	for _, denom := range bep3Denoms {
```
Comment (nddeluca) [inline x/evmutil/keeper/conversion_evm_native_bep3_test.go:28]: This actually makes the tests quite a bit easier to read -- easy to see that all denoms are tested and much more resistant to refactorings that could drop a denom, accidentally rename one, break the check some other unique way, etc

File: x/evmutil/keeper/conversion_evm_native.go [hunk 8]
```diff
@@ -102,13 +110,23 @@ func (k Keeper) ConvertERC20ToCoin(
 		return err
 	}
 
+	amountToLock := amount.BigInt()
+	amountToMint := amount.BigInt()
+
+	if isBep3Asset(pair.Denom) {
```
Comment (nddeluca) [inline x/evmutil/keeper/conversion_evm_native.go:116]: I see what you mean know, and this is a nice approach showing there is are two paths in the logic: One for bep3 assets, and one for not.  Also, makes it clear that the amounts are only modified in the bep3 case, but also keeps the dirty details down a level in a function call.

**Review by nddeluca — APPROVED** [review]

File: x/evmutil/keeper/conversion_evm_native_bep3.go (hunk 5 above)
Comment (DracoLi) [reply x/evmutil/keeper/conversion_evm_native_bep3.go:27]: good point, i think this will benefit here
